derive/src/lib.rs

```diff
@@ -67,21 +77,51 @@ fn add_trait_bounds(mut generics: Generics, lifetime: LifetimeDef) -> Generics {
     generics
 }
 
-fn gen_arbitrary_method(input: &DeriveInput, lifetime: LifetimeDef) -> TokenStream {
+fn with_recursive_count_guard(
+    recursive_count: &syn::Ident,
+    expr: impl quote::ToTokens,
+) -> impl quote::ToTokens {
+    quote! {
+        #recursive_count.with(|count| {
+            if count.get() > 0 && u.is_empty() {
+                return Err(arbitrary::Error::NotEnoughData);
+            }
+
+            count.set(count.get() + 1);
+            let result = { #expr };
+            count.set(count.get() - 1);
```

Comment:
Does this work as intended, i.e. is it ok to skip decreasing count if `#expr` aborts through `?`?

<details>
<summary>Generated code for my struct, for consideration</summary>

```rust
fn arbitrary(u: &mut Unstructured) -> Result<Self> {
    RECURSIVE_COUNT_Nah.with(|count| {
        if count.get() > 0 && u.is_empty() {
            return Err(arbitrary::Error::NotEnoughData);
        }
        count.set(count.get() + 1);
        let result = {
            Ok(
                match (u64::from(<u32 as Arbitrary>::arbitrary(u)?) * 2u64) >> 32 {
                    0u64 => Nah::Foo(
                        Arbitrary::arbitrary(u)?,
                        Arbitrary::arbitrary(u)?,
                        Arbitrary::arbitrary(u)?,
                    ),
                    1u64 => Nah::Bar(Arbitrary::arbitrary(u)?, Arbitrary::arbitrary(u)?),
                    _ => panic!("internal error: entered unreachable code"),
                },
            )
        };
        count.set(count.get() - 1);
        result
    })
}
```

</details>

Reply:
Actually, thinking about it further, wouldn't it be better to only modify `count` when the unstructured is already empty?
* No need to access the thread local under normal conditions (performance micro-optimization, I know.)
* It's entirely possible that a recursive struct is benign and does finish generating on zeros only. But this currently aborts before trying that.